### PR TITLE
Fix graphical glitches on Windows

### PR DESCRIPTION
**Pull request by rustbasic**
Fix graphical glitches on Windows

request_repaint()
if is_minimized return

Issues :
#3321

### CODE REVIEW

**Comment by rustbasic** [comment]
@emilk 
Dear emilk.

Pull Request.

Based on long-term testing on Windows, ignoring request_repaint() when minimized prevents graphical corruption.

Since `i.viewport().minimized` is `true` only on Windows,
I don't think we need to do `cfg!(target_os = "windows")`

```
if cfg!(target_os = "windows") && self.input(|i| i.viewport().minimized.unwrap_or(false)) {
    return;
}
```

**Comment by valaphee** [comment]
but with cfg, the if gets optimized out as a result

**Comment by rustbasic** [comment]
> but with cfg, the if gets optimized out as a result

Even in non-windows, there is no need to run request_repaint if it is minimized.

File: crates/egui/src/context.rs [hunk 1]
```diff
@@ -1240,6 +1240,9 @@ impl Context {
     ///
     /// This will repaint the current viewport.
     pub fn request_repaint(&self) {
+        if self.input(|i| i.viewport().minimized.unwrap_or(false)) {
+            return;
```
Comment (emilk) [inline crates/egui/src/context.rs:1244]: I don't think this is the right solution.

The core problem is that when minimized, `App::update` is being called with a bad viewport size. We should either make sure `App::update` isn't called when minimized, or if called it is done so with the size of the window from when it was large.

This fix will just fix _some_ reasons for why `App::update` might be called, but not all. For instance, calling `request_repaint_after`, and then minimizing, will still call `App::update`.

TL;DR: the fix needs to into `eframe`, not `egui`

**Review by emilk — CHANGES_REQUESTED** [review]

File: crates/egui/src/context.rs (hunk 1 above)
Comment (rustbasic) [reply crates/egui/src/context.rs:1244]: @emilk
Dear emilk,

I'm not sure if this is the right solution, but I've been able to avoid the problem with this method after testing it for about a year.

I think App::update() should continue to work and should not only output to the screen.

In most cases, most users only call request_repaint(), and there are no cases where request_repaint_after() is called.

Please make a Pull Request for this temporarily until I find a better way. 
If there is a better way, please fix it.

Thank you, emilk.

Comment (rustbasic) [reply crates/egui/src/context.rs:1244]: If you want, let it return only on Windows.

File: crates/egui/src/text_selection/text_cursor_state.rs
Comment (rustbasic) [inline crates/egui/src/text_selection/text_cursor_state.rs:1]: rev() is panic with UTF8

File: crates/egui/src/widgets/text_edit/text_buffer.rs
Comment (rustbasic) [inline crates/egui/src/widgets/text_edit/text_buffer.rs:1]: Not byte_index
It's char_index

#3846

File: crates/egui/src/text_selection/text_cursor_state.rs
Comment (rustbasic) [inline crates/egui/src/text_selection/text_cursor_state.rs:1]: @emilk 
Dear emilk.

I want to use the following for accurate usage:                                       
`let width = unicode_width::UnicodeWidthChar::width_cjk(c).unwrap_or(2);`             

Thank you, emilk.

**Comment by rustbasic** [comment]
@emilk 
Dear emilk.

I work finished, Pull Request.
#3846 
#3850

I want to use the following for accurate usage:
egui\src\text_selection\text_cursor_state.rs 360 lines
`let width = unicode_width::UnicodeWidthChar::width_cjk(c).unwrap_or(2);`

Thank you, emilk.

File: crates/egui/src/context.rs (hunk 1 above)
Comment (rustbasic) [reply crates/egui/src/context.rs]: Changed :
crates/eframe/src/native/run.rs 

The test results are normal.
